### ISSUE TITLE
Ability to import project into Eclipse automatically

### ISSUE THREAD

**Issue by hohwille** [devonfw/ide#175]
For Eclipse (and with additional issues also for other IDEs) we should introduce the ability to import an existing (maven) project into the current workspace:
```
devon eclipse import [«path»]
```
Should import the specified project from the given `«path»` into the current Eclipse workspace. If `«path»` is omitted, the current working directory is used.
My proposal would be to always use the current workspace:
```
cd workspaces/main
devon eclipse import ../test-workspace/myproject
```
Would import `myproject` into the Eclipse of the `main` workspace and not into `test-workspace`, what IMHO is consistent and allows special cases (e.g. importing `workspaces/main/CobiGen_Templates` into all workspaces using `--all` option).

If the project is a maven project, the import should behave like `Import existing maven projects into workspace`. Other cases are just nice-to-have (gradle, general plain project).
IMHO the most tricky thing is to figure out, how to trigger the import in Eclipse from CLI.

As extension it should also be possible to provide a git URL for `«path»`. This would clone the git URL to the current working directory and import the resulting project.

If we managed to do that, we should also extend `java` commandlet such that it could clone and import my-thai-star into the current workspace (unless already done).

**Comment by hohwille** [comment]
https://www.eclipse.org/forums/index.php/t/109302/

**Comment by hohwille** [comment]
As an alternative we could do that manually once and observe all that Eclipse is doing:
* create `.project`
* create `.classpath`
* create changes in workspace `.metadata` folder

This is however reinvent the wheel and would be hard to maintain. If we are able to trigger the eclipse feature programatically it would be awesome.

**Comment by hohwille** [comment]
I tried `eclipse -application org.eclipse.cdt.managedbuilder.core.headlessbuild -import ...` but Eclipse is a big mess as always:
```
Could not resolve module: org.eclipse.mylyn.bugzilla.core [277]
  Unresolved requirement: Require-Bundle: org.apache.xmlrpc
    -> Bundle-SymbolicName: org.apache.xmlrpc; bundle-version="3.0.0.v20100427-1100"
```
I "resolved" this by uninstalling bugzilla support for the moment and ran into the next issue:
```
Could not resolve module: org.eclipse.epp.logging.aeri.ide [145]
  Unresolved requirement: Import-Package: org.apache.lucene.document; version="[7.1.0,8.0.0)"
```
This lead me to this bug: https://bugs.eclipse.org/bugs/show_bug.cgi?id=540306
I added the nightly update site and tried to install the update what again failed.
Then retried to just install the `EEP Logging Third-Party Dependencies` what lead also to several other updates and then worked.
Now I get:
```
SESSION 2019-08-19 12:33:03.746 -----------------------------------------------
eclipse.buildId=4.12.0.I20190605-1800
java.version=11.0.4
java.vendor=AdoptOpenJDK
BootLoader constants: OS=macosx, ARCH=x86_64, WS=cocoa, NL=en_DE
Framework arguments:  -product org.eclipse.epp.package.java.product -application org.eclipse.cdt.managedbuilder.core.headlessbuild -import /projects/ide-test/workspaces/main/keywi
Command-line arguments:  -os macosx -ws cocoa -arch x86_64 -product org.eclipse.epp.package.java.product -data /projects/ide-test/workspaces/main -application org.eclipse.cdt.managedbuilder.core.headlessbuild -import /projects/ide-test/workspaces/main/keywi

!ENTRY org.eclipse.osgi 4 0 2019-08-19 12:33:04.573
!MESSAGE Application error
!STACK 1
```
So the plugin errors are gone but funny thing is that headlessbuild import fails with Application error without any further details.
Seems as if this feature of eclipse is unmaintained and entirely broken. Pitty.

**Comment by hohwille** [comment]
https://bugs.eclipse.org/bugs/show_bug.cgi?id=535768

**Comment by hohwille** [comment]
I tried the OOMPH Project Setup.
Update Site:
http://download.eclipse.org/oomph/updates

Tried this config:
```
ECLIPSE_VMARGS=-Doomph.redirection.index.redirection=index:/org.eclipse.setup->file:///projects/ide-test/org.eclipse.setup -Doomph.update.url=http://download.eclipse.org/oomph/updates/milestone/latest
```
In the specified location I put a modified copy of http://devonfw.github.io/devon-ide/oomph/index/org.eclipse.setup

Now when I launch eclipse with this options I can verify that these system properties for OOMPH have been set (in help). However, nothing happens at all - no effect. I have not a single clue how to debug or what else to try.
Seems I need some OOMPH expert to get anything working with this OOMPH project setup...

**Comment by phschimm** [comment]
Based on @maybeec's [oomph poject setups](https://github.com/devonfw/devon-ide/tree/gh-pages/oomph/projects), @dhelmke and I have created a very basic project setup that can clone any arbitrary GitHub repo (specified by the user) and will then proceed to import all Maven projects contained in the `/java` sub-directory of that repo:

> [GitHubCloneMavenImport.setup](https://github.com/dhelmke/NewAsciidocFiles/blob/master/documentation/GitHubCloneMavenImport.setup)

We are now looking into the possibility of invoking this process from the command line.
I suspect we'll have to include an _Eclipse Installer_ executable in the software shipped with devonfw to pull that off ...

**Comment by maybeec** [comment]
the eclipse installer is an own eclipse distribution. Everthing which is in there especially the execution code for oomph profiles is already available in every eclipse. So the question is more how let the available installation done by devon-ide take up the profile.

First of all you have to put your oomph setup to some place including the indexes and so on as you can see here: http://devonfw.github.io/devon-ide/oomph/index/org.eclipse.setup

Then call your eclipse with an appropriate redirect to this folder by adding `-Doomph.redirection.index.redirection=index:/org.eclipse.setup->http://devonfw.github.io/devon-ide/oomph/index/org.eclipse.setup` but just with a local path.

Then the magic path for including it automatically to the workspace is `.metadata/.plugins/org.eclipse.oomph.setup/workspace.setup`, where you can refer the projects being checked out in a project. So this file specifies the reference to a project setup for example and its variable instantiations to run it automatically during eclipse startup.

**Comment by phschimm** [comment]
Thank you @maybeec, that put us on the right track.

We can simply manipulate the `workspace.setup` file in `/.metadata/.plugins/org.eclipse.oomph.setup/` via bash to include a _MavenImportTask_ (and optionally a _GitCloneTask_), and our Eclipse instance will automatically import all Maven projects contained in that workspace folder. The .setup file would have to look somewhat like this:

```
<?xml version="1.0" encoding="UTF-8"?>
<setup:Workspace
    xmi:version="2.0"
    xmlns:xmi="http://www.omg.org/XMI"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:maven="http://www.eclipse.org/oomph/setup/maven/1.0"
    xmlns:setup="http://www.eclipse.org/oomph/setup/1.0"
    xsi:schemaLocation="http://www.eclipse.org/oomph/setup/maven/1.0 http://git.eclipse.org/c/oomph/org.eclipse.oomph.git/plain/setups/models/Maven.ecore"
    name="workspace">
<setupTask
    xsi:type="maven:MavenImportTask">
    <sourceLocator
        rootFolder="${WORKSPACE_PATH}"
        locateNestedProjects="true"/>
</setupTask>
</setup:Workspace>
```

After the inital import we have to revert the changes to the `workspace.setup` file, otherwise Eclipse will run the _MavenImportTask_ at every startup of that instance (which is slow, especially when `locateNestedProjects` is set to true).

We can decide between two approaches when it comes to the _Git clone_ workflow:
1. The user passes his desired Git-URL via command line and we insert it into `workspace.setup` before the startup of our Eclipse instance.
2. We set the Git-URL as an undefined variable in our `workspace.setup`. This causes the Eclipse instance to open an input-prompt on launch, into which the user can then paste the URL.

**Comment by phschimm** [comment]
[Here's](https://github.com/phschimm/devon-ide/blob/7ecd8c5b9f80c965595de1fcb111ae5c3219ecde/scripts/src/main/resources/scripts/command/eclipse#L167) my inital implementation of a `devon eclipse mvn-import` command as a precursor to a more universal/complex `devon eclipse import` command. This command utilizes a template file located in the newly created directory `/scripts/templates/oomph/`, akin to the jenkins template approach by @hohwille (thanks again for the hint).
I'm slowly getting a grip on Bash scripting and will be adding more functionality in the coming days. Currently I have a problem however:

I'm looking for a way to execute some function in my script, when a process — which I invoked from that script beforehand — terminates.

Does anyone have an idea how this could be done?

**Comment by phschimm** [comment]
I created [pull request #247](https://github.com/devonfw/ide/pull/247) for my implementation.
I would be thankful if someone could test this under Mac/Linux.

**Comment by markusschuh** [comment]
I have tried the current state (`eclipse-import` branch) and succeeded to import my-thai-star after a few tries. So basically the automated import works already - good work!

Some remaining challenges that may be worth to address within the ongoing work. This isn't any criticism - just feedback about my current findings:

1. There is no result returned yet from the "oomph setup" that the `devon eclipse` commandlet would check I don't know, if this is possible at all - but it would at least help, if the the script would know, if the "oomph setup" has worked - or not.
1. I have needed several tries until the first access. This was because some preconditions had first to be installed. On first run, the `workspace.setup` was still missing - but `devon eclipse import ...` had reported what need to be done => good.  Then the eclipse wanted to update the "oomph setup" and restart eclipse. Also o.k. - but it was a bit irritating, when the `devon eclipse import ...` finally reported "CLEANUP SUCCESSFUL!" - and nothing had happened. And when I later have tried the git import again some additional component had to be installed  - and the setup repeated.
1. I had difficulties to understand, what git URL had to be used. On first try, I have missed the final `.git`. After code inspection, I have added it - but still I miss, how I could specify another branch. These details should be documented, if not yet realized.
1. Finally the "oomph setup" asked me for a missing but mandatory variable: `scope.project.stream.name` I entered a dummy value  - and import succeeded. I assume, this might just be caused by the enforced install of a newer oomph version - so this run may have used a version, you have not yet used. 

In summary nothing that could not be solved Nevertheless as long as the need to address all this challenges is there, the advantage of this CLI based import vs. the standard import via Eclipse UI is not yet that high.  But I guess, this will change, since this feature / branch is yet unfinished.

**Comment by phschimm** [comment]
Hi @markusschuh, thank you for the feedback.
I recently met with @hohwille to discuss the code and we came to a similar conclusion as you did in points 1. and 2.: The current workflow is unintuitive, cumbersome and error-prone. This is a result of our current approach; we don't really interact with Eclipse/Oomph but are only replacing a file, which is then read on startup. We have decided to scrap this approach and focus on developing an Eclipse plug-in instead, which will allow us to interact with Eclipse/Oomph more closely.

Points 3. and 4. are actually related; the requested `scope.project.stream.name` is basically the branch that Oomph will check out during cloning. I didn't encounter this issue during my testing, probably because I defined that variable in an early test once and it was set on my system from there on out. But it makes a lot of sense, now that you bring it up.

**TL;DR:** We are now approaching this feature differently by working on an Eclipse plug-in. As a result, the `eclipse-import` working branch should probably be deleted.

**Comment by hohwille** [comment]
I want to do a fresh start on this:
1. projects should be able to preconfigure projects that will be cloned and potentially imported into IDEs
2. This feature should be a general feature not tightly coupled to Eclipse hence OOMPH may be helpful to import the project(s) into Eclipse but the process (e.g. cloning) should be in a `devonfw-ide` commandlet itself.
3. The configuration via `devon.properties` might be tedious. We could use an array variable but a dedicated config file (e.g. JSON) would be better. I do not expect that users need any kind of overriding features for this.
4. For each project we need to define `scmURL` (initially only git is supported), `scmBranch` (default branch if omitted), `workspace` (`main` by default), `ide` (`eclipse` to import the project into Eclipse, later extendable as comma separated list also supporting IntelliJ, etc.).
5. During the setup `devonfw-ide` will go through all these configured projects: If they already exist in the configured workspace, they get updated (may also be configurable to switch off this behavious) otherwise they will get cloned in this location and imported into configured the IDE(s).
6. The functionality to import a project into Eclipse should be a general feature of `eclipse` commandlet that is only used here but may also be used manually via CLI

I hope this all makes sense. Feedback is welcome.

**Comment by hohwille** [comment]
In case we are trying to implement 6. using OOMPH I would try to build some kind of "queue" as a simple flat file(s) where each request to import a file is written to. When eclipse is started, all "queued" import requests shall be passed to OOMPH as tasks for automatic execution and these entries need to be removed from the "queue".
OOMPH is not properly designed for such a queue, but we could simply create a folder inside `DEVON_IDE_HOME` where we put a single file for each import request. Next we extend our `eclipse` commandlet to check this "queue" folder. If it is not empty, all entries shall be transformed into a single OOMPH XML file and then be deleted. Then eclipse is launched with the required parameters so this XML is read. As files and processes like this are not transactional in any way, we can not easily know if OOMPH has processed the XML. Therefore we should not override it. It could therefore contain a timestamp to prevent override. Further, we could add a last task in our logic that will delete the XML file from the disc. This way we could figure out, that OOMPH has completely processed the file. If we are in the logic to lauch eclipse and read the "queue" we could also honor XML files that have not been processed and requeue them.
This makes the following idea:
Our basedirectory for the "queue" will be `workspace/«name»/.metadata/.devon-ide/imports` with:
* `queue` as a subfolder where we write import requests as `«timestamp».oomph`. The content would be the XML fragment with all tasks for OOMPH to be added to the final OOMPH XML.
* `work` as a subfolder where the aggregated content goes. When Eclipse is launched, then a new `work/«timestamp»` folder is created as subfolder here. A `work/«timestamp»/queue` folder is created as subfolder and all files from `queue` are processed and then moved to this new folder. Processing will be easy: Initially do `echo "«XML HEADER»" > work/«timestamp»/oomph.xml` and for every file do `cat «file» >> work/«timestamp»/oomph.xml`. Also do the same for all other `work/«timestamp»/queue` subfolders and delete those `work/«timestamp»` folders when processed. Finally do `echo "«XML FOOTER»" > work/«timestamp»/oomph.xml`. This footer will contain the OOMPH task to delte the current `work/«timestamp»` folder and the closing root tag.
* Finally Eclipse is launched with the `workspace/«name»/.metadata/.devon-ide/imports/work/«timestamp»/oomph.xml` as param for OOMPH.

**Comment by hohwille** [comment]
Fixed by the linked and merged PRs.
